### PR TITLE
feat(scrapers.tasks): Add update documents from text function to cron

### PR DESCRIPTION
**Pull request by flooie**
update_document_from_text adds cluster and citation information from
scraped pdfs, currently only tax court is supported

### CODE REVIEW

File: cl/scrapers/tasks.py [hunk 1]
```diff
@@ -228,6 +231,32 @@ def get_page_count(path, extension):
...
+            if mdl.__name__ == "OpinionCluster":
+                obj = mdl.objects.get(pk=opinion.cluster_id)
+                obj.__dict__.update(**metadata_dict[mdl.__name__])
+            elif mdl.__name__ == "Citation":
+                metadata_dict[mdl.__name__]["cluster_id"] = opinion.cluster_id
+                mdl.objects.get_or_create(**metadata_dict[mdl.__name__])
+            else:
+                raise NotImplemented(
+                    "Object type of %s not yet supported." % mdl.__name__
+                )
```
Comment (mlissner) [inline cl/scrapers/tasks.py]: OK, I don't know if the following will work, but I think it's a good simplification of what you've got. Why I like it:

1. The main benefit here is that we don't have two large variables we're working with. In your version, we're looking iterating over `models` and then looking stuff up in `metadata_dict` according to what we get in `models`. In my version, we eliminate `models`, iterate over `metadata_dict`, and do no lookups. 

2. It's shorter (at least until it's blackified, sigh)

3. I think it's more readable?

4. It gets the cluster in a much simpler way, which prevents a DB lookup. The cluster gets saved later on, so this should be fine, I think. 

```suggestion
for model_name, data in metadata_dict.items():
    ModelClass = apps.get_model("search.%s" % model_name)
    if model_name == "OpinionCluster":
        cluster = opinion.cluster.__dict__.update(data)    
    elif model_name == "Citation":
        ModelClass.objects.get_or_create(**data)
    else:
        raise NotImplemented("Object type of %s not yet supported." % model_name)
```

Anyway, this took me a few minutes of staring to figure out, and I haven't tested it, but it does feel more legible to me. (Your version, if we stick with it, would also benefit a lot by pulling `mdl.__name__` into a variable.)

Comment (flooie) [reply cl/scrapers/tasks.py]: Does this need to be stored in the cluster 
```cluster = opinion.cluster.__dict__.update(data)  ```

Comment (flooie) [reply cl/scrapers/tasks.py]: we still need to set the cluster id for citation generation
```data["cluster_id"] = op.cluster_id```

Comment (flooie) [reply cl/scrapers/tasks.py]: I think this works
```     
for model_name, data in cd.items():
      ModelClass = apps.get_model("search.%s" % model_name)
      if model_name == "OpinionCluster":
          op.cluster.__dict__.update(**data)
      elif model_name == "Citation":
          data["cluster_id"] = op.cluster_id
          ModelClass.objects.get_or_create(**data)
      else:
          raise NotImplemented("Object type of %s not yet supported." % model_name```

**Comment by flooie** [comment]
I think this is ready for review again.

File: cl/lib/juriscraper_utils.py [hunk 2]
```diff
@@ -0,0 +1,21 @@
+import pkgutil
+import juriscraper
+import importlib
+
+
+def get_scraper_object_by_name(court_id):
+    """Identify and instantiate a Site() object given the name of a court
+    :param court_id: The ID of a site; must correspond with the name of a
```
Comment (mlissner) [inline cl/lib/juriscraper_utils.py]: Missing a newline between summary and params.

File: cl/scrapers/tasks.py [hunk 3]
```diff
@@ -228,6 +231,30 @@ def get_page_count(path, extension):
...
 
 
+def update_document_from_text(opinion):
+    """
+    Given opinion object, generate Juriscraper site object and
+    extract metadata from text (currently only implemented in Tax
+
+    :param opinion: Opinion Object
+    :return:
+    """
```
Comment (mlissner) [inline cl/scrapers/tasks.py]: Missing paren, but also, this should have a one line summary and then if needed a multi-line description after a line return. 

Notionally:

```suggestion
    """Extract additional metadata from document text
    
    Use functions from Juriscraper to pull metadata out of opinion
    text. Currently only implemented in Tax, but functional in all 
    scrapers via AbstractSite object.

    :param opinion: Opinion object
    :return: None
    """
```

**Comment by mlissner** [comment]
Also, one other thing, not sure why, but tests are failing.

File: cl/scrapers/tasks.py (hunk 3 above)
Comment (flooie) [reply cl/scrapers/tasks.py]: It could be failing because the latest juriscraper master has not yet been pushed.  The two are being updated side by side.

**Comment by mlissner** [comment]
> It could be failing because the latest juriscraper master has not yet been pushed. The two are being updated side by side.

Right. I guess we're ready for me to deploy a new version to docker aren't we?

File: cl/scrapers/tasks.py (hunk 1 above)
Comment (flooie) [reply cl/scrapers/tasks.py]: I think this has been resolved

**Comment by flooie** [comment]
This will continue to fail until we merge https://github.com/freelawproject/juriscraper/pull/309 and push a new version to PyPi and docker.  Everything is functional in my dev environment.

File: cl/scrapers/tasks.py [hunk 4]
```diff
@@ -248,6 +248,7 @@ def update_document_from_text(opinion):
             opinion.cluster.__dict__.update(data)
         elif model_name == "Docket":
             opinion.cluster.docket.__dict__.update(data)
+            opinion.cluster.docket.save()
```
Comment (flooie) [inline cl/scrapers/tasks.py]: I was hesitant to add this line, but when I either added it later in the function where cluster and opinion are saved my tests failed.

File: cl/scrapers/tests.py [hunk 5]
```diff
@@ -124,22 +124,21 @@ def test_juriscraper_object_creation(self):
...
+            None,
+            o.cluster.docket.docket_number,
+            msg="Docket number should be none.",
+        )
+        extract_doc_content(pk=76, do_ocr=False)
+        o = Opinion.objects.get(pk=76)
+        self.assertEqual(
+            "19031-13, 27735-13, 11905-14",
+            o.cluster.docket.docket_number,
+            msg="Docket number extracted properly",
```
Comment (flooie) [inline cl/scrapers/tests.py]: I had a duplicated test - I assume during a rebase.  I replaced the duped test to test docket number extraction.  Tests fail currently, but with the PR juriscraper this test succeeds.

File: cl/scrapers/tasks.py (hunk 3 above)
Comment (flooie) [reply cl/scrapers/tasks.py]: Fixed and updated

File: cl/scrapers/tests.py (hunk 5 above)
Comment (flooie) [reply cl/scrapers/tests.py]: This has been fixed and all tests are now passing with the docker image being pushed

**Comment by mlissner** [comment]
Code is all good, thanks. One thing that's still messy is the git history. We'll want to work on cleaning that up in future PRs, but for this one I'm going to squash everything during the merge and that'll take care of it well enough. But this PR is +25, -6 lines and 24 commits.

**Pull request merged**